go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,544 @@
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// maxRetries is the number of times a object will be retried before it is dropped out of the queue.
+	// With the current rate-limiter in use (5ms*2^(maxRetries-1)) the following numbers represent the
+	// sequence of delays between successive queuings of an object.
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	// maxRetries = 15
+
+	controllerName  = "net-attach-def-controller"
+	avoidResync     = 0
+	numberOfWorkers = 2
+	qps             = 15
+	maxRetries      = 10
+)
+
+type BaseNetworkController interface {
+	Start(ctx context.Context) error
+	Stop()
+	GetNetworkName() string
+}
+
+type NetworkController interface {
+	BaseNetworkController
+	CompareNetConf(util.NetConfInfo) bool
+	AddNad(nadName string)
+	DeleteNad(nadName string)
+	IsNadExist(nadName string) bool
+	// Cleanup cleans up the given network, it could be called to clean up network controllers that are deleted when
+	// ovn-k8s is down; so it's receiver could be a dummy network controller.
+	Cleanup(netName string) error
+}
+
+// NetworkControllerManager manages all network controllers
+type NetworkControllerManager interface {
+	NewNetworkController(netInfo util.NetInfo, netConfInfo util.NetConfInfo) (NetworkController, error)
+	SyncNetworkControllers(allControllers []NetworkController) error
+	// NewDummyNetworkController creates a dummy network controller used to clean up specific network
+	NewDummyNetworkController(topologyType, netName string) (NetworkController, error)
+}
+
+type netAttachDefinitionController struct {
+	ncm                NetworkControllerManager
+	nadFactory         networkattachmentdefinitioninformerfactory.SharedInformerFactory
+	netAttachDefLister networkattachmentdefinitionlisters.NetworkAttachmentDefinitionLister
+	netAttachDefSynced cache.InformerSynced
+	queue              workqueue.RateLimitingInterface
+	loopPeriod         time.Duration
+
+	NetAttachDefManager
+}
+
+func NewNadController(ncm NetworkControllerManager, ovnClientset *util.OVNClientset) *netAttachDefinitionController {
+	nadFactory := netattachdefinformers.NewSharedInformerFactoryWithOptions(
+		ovnClientset.NetworkAttchDefClient,
+		avoidResync,
+	)
+	netAttachDefInformer := nadFactory.K8sCniCncfIo().V1().NetworkAttachmentDefinitions()
+	rateLimter := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(qps), qps*5)})
+
+	nadController := &netAttachDefinitionController{
+		ncm:                ncm,
+		nadFactory:         nadFactory,
+		netAttachDefLister: netAttachDefInformer.Lister(),
+		netAttachDefSynced: netAttachDefInformer.Informer().HasSynced,
+		queue:              workqueue.NewNamedRateLimitingQueue(rateLimter, "net-attach-def"),
+		loopPeriod:         time.Second,
```

Comment:
can you elaborate on why a second ? Was this copied from somewhere else ?

Reply:
you had this framework done, and I just used your code and didn't change it. do you suggest other value?